### PR TITLE
[RLlib] Actually save the optimizer state for tf learners

### PR DESCRIPTION
**Pull request by avnishn**
It turns out you can get the actual optimizer state by calling optimizer.variables for tf keras.
this pr enables us to save the full optimizer state and restore it. To do this I added a new
file called optimizer_name_state.txt to the checkpoint. This holds a bytestring serialized
representation of the optimizer's state. It looks like the optimizer's variable state doesn't include
things like the learning rate, so I still need to save those as a separate file and
reconstruct the optimizer first before loading the state.

Signed-off-by: Avnish <avnishnarayan@gmail.com>

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by kouroshHakha — APPROVED** [review]
sounds good. If you can just polish this a little bit so that it looks more modular.

**Comment by avnishn** [comment]
broken tests are unrelated. 

the broken doc test is addressed here:
https://github.com/ray-project/ray/pull/34291

**Comment by avnishn** [comment]
failing learning tests are not impacted by this pr since they are not currently on the new learner stack
